src/os/ui/util/punyparent.js

```diff
@@ -29,95 +30,184 @@ os.ui.Module.directive('punyparent', [os.ui.util.punyParentDirective]);
+   * @type {Function}
+   * @private
+   */
+  this.updateCollapsedFn_ = this.updateCollapsed_.bind(this);
+
+  /**
+   * Throttle to limit how often the controller responds to viewport size changes.
+   * @type {goog.Throttle}
+   * @private
+   */
+  this.vsmThrottle_ = new goog.Throttle(this.updateCollapsed_, 200, this);
+
+  // Listen for viewport size changes.
+  const vsm = goog.dom.ViewportSizeMonitor.getInstanceForWindow();
+  vsm.listen(goog.events.EventType.RESIZE, this.vsmThrottle_.fire, false, this.vsmThrottle_);
+
+  // Update the collapsed state when the root element size changes.
+  os.ui.resize(this.element_, this.updateCollapsedFn_);
 };
 
 
 /**
- * @private
+ * Angular $onDestroy lifecycle hook.
  */
-os.ui.util.PunyParentCtrl.prototype.destroy_ = function() {
-  var vsm = goog.dom.ViewportSizeMonitor.getInstanceForWindow();
-  vsm.unlisten(goog.events.EventType.RESIZE, this.onResize_, false, this);
+os.ui.util.PunyParentCtrl.prototype.$onDestroy = function() {
+  os.ui.removeResize(this.element_, this.updateCollapsedFn_);
+
+  this.observer_.disconnect();
+
+  this.watchedChildren_.forEach((child) => {
+    os.ui.removeResize($(child), this.onChildResizeFn_);
+  });
+  this.watchedChildren_.length = 0;
+
+  goog.dispose(this.childResizeDelay_);
+  this.childResizeDelay_ = null;
+
+  const vsm = goog.dom.ViewportSizeMonitor.getInstanceForWindow();
+  vsm.unlisten(goog.events.EventType.RESIZE, this.vsmThrottle_.fire, false, this.vsmThrottle_);
+
+  goog.dispose(this.vsmThrottle_);
+  this.vsmThrottle_ = null;
 
-  if (this.throttle_) {
-    this.throttle_.dispose();
-    this.throttle_ = null;
-  }
   this.scope_ = null;
   this.element_ = null;
 };
 
 
 /**
+ * Handle mutation event.
+ * @param {!Array<!MutationRecord>} mutationsList The mutation list.
+ * @param {!MutationObserver} observer The observer.
  * @private
  */
-os.ui.util.PunyParentCtrl.prototype.onResize_ = function() {
-  this.throttle_.fire();
+os.ui.util.PunyParentCtrl.prototype.onMutation_ = function(mutationsList, observer) {
+  for (const mutation of mutationsList) {
+    if (mutation.type === 'childList') {
+      mutation.addedNodes.forEach(this.watchChild_, this);
+      mutation.removedNodes.forEach(this.unwatchChild_, this);
+    }
+  }
 };
 
 
 /**
+ * Watch a child node.
+ * @param {!Node} child The child.
  * @private
  */
-os.ui.util.PunyParentCtrl.prototype.onThrottleResize_ = function() {
-  if (this.element_) {
-    var children = this.element_.children().toArray();
-    var childrenWidth = children.reduce((accumulator, child) => {
-      var c = $(child);
-      // ignore the resize trigger since thats the parent size
-      if (!c.hasClass('resize-triggers')) {
-        accumulator += c.outerWidth(true);
-      }
-      return accumulator;
-    }, 0);
+os.ui.util.PunyParentCtrl.prototype.watchChild_ = function(child) {
+  const childEl = $(child);
+  if (!childEl.hasClass('resize-triggers') && !childEl.hasClass('resize-sensor')) {
```

Comment:
Are these class exclusions to ensure that we don't double-watch a child somehow? Is there a concern where something might have its own internal resize listener that adds these, causing us miss watching it here?

Reply:
The resize libraries add elements to the DOM that they use to determine when a resize happens. These elements have absolute positioning and 100% width so their width is the size of the entire nav bar, but they aren't visible and don't contribute to determining the relative size of things in the nav. We want to ignore them, and I had to add the `resize-sensor` class because it's used by the new resize library that backs `os.ui.resize`. I think `resize-triggers` was the class for the old resize library.

Reply:
Ahh gotcha, I thought the classes were added to the existing elements being checked for resize rather than entirely new dummy elements.